### ISSUE TITLE
Change optimizer after a phase transition

### ISSUE THREAD

**Issue by samgelman** [speediedan/finetuning-scheduler#6]
## 🚀 Feature

Hello! I would like to change the optimizer after a phase transition, similar to [changing the learning rate schedule after a phase transition](https://github.com/speediedan/finetuning-scheduler/issues/1). Is this possible? If not, would it be possible to implement as a new feature for finetuning-scheduler?

### Motivation

I want to first train a linear top net with SGD then switch to Adam to finetune the whole network. I found that Adam does not converge to the optimal solution for my linear top nets, so that's why I want to use SGD for the first phase. 

### Pitch

It would be great if we could change the optimizer after a phase transition, similar to changing the learning rate scheduler. We should be able to change both optimizer and learning rate scheduler together. 

### Additional context

______________________________________________________________________

**Comment by speediedan** [comment]
Thanks for taking the time to file the feature request @samgelman! You're not alone in requesting this feature and I certainly plan to add it in the not-too-distant future. 
 
I've been following PyTorch Lightning's requested implementation of [this feature](https://github.com/Lightning-AI/lightning/issues/12712) to facilitate FTS's support for per-phase optimizer specification. That feature would make implementation of per-phase optimizer re-instantiation notably cleaner, straightforward and robust I suspect. As soon as I have a chance to connect with the Lightning team regarding their plans and/or expected implementation timeline for [#12712](https://github.com/Lightning-AI/lightning/issues/12712), I'll decide whether to attempt implementation of this feature without that PL functionality. My current thinking is if [#12712](https://github.com/Lightning-AI/lightning/issues/12712) isn't on the roadmap for PL 1.9 and my time permits, I'll likely attempt implementation and update this issue accordingly. 
 
Of course, PRs are welcome anytime as well if you or anyone else is interested in tackling it in the meantime! 😀
Thanks again for proposing this feature and reaching out. Feel free to reach out anytime if you have other issues or want to share your use case. Best of luck with your work!

**Comment by speediedan** [comment]
Thanks again for your valuable feedback @samgelman! 

I'm closing this as the [requested feature](https://finetuning-scheduler.readthedocs.io/en/stable/advanced/optimizer_reinitialization.html) is now available (in beta) with Fine-Tuning Scheduler [2.0.2](https://github.com/speediedan/finetuning-scheduler/releases/tag/v2.0.2). Apologies this one took a while, it kept getting deferred waiting for the aforementioned upstream changes but I ultimately decided it could be implemented sufficiently cleanly without said changes. Hope it works for you and/or others. Good luck with your research!
